### ISSUE TITLE
Evaluation Code Bugs

### ISSUE THREAD

**Issue by ricky-696** [salmank255/ROAD_Waymo_Baseline#11]
Hi @salmank255,

First, I encountered an issue where the agent scores were all 0 points. After tracing `evaluation.py`, I identified three bugs. I hope you can address them(Track1 and Track2 are the same):

The first issue causing 0 scores:

1. In [line 276 of dataset.py](https://github.com/salmank255/ROAD_Waymo_Baseline/blob/65049d6e633c90cb5924b36a4c74f69e5551d491/data/datasets.py#L276), I noticed that the default values for `g_h` and `g_w` are `0` in [line 27](https://github.com/salmank255/ROAD_Waymo_Baseline/blob/65049d6e633c90cb5924b36a4c74f69e5551d491/data/datasets.py#L27), but the `get_filtered_tubes` function does not receive `g_h` and `g_w` as parameters. This leads to ground truth calculations resulting in an IoU of `0`.

Next, the problem of why participants' scores (including the baseline) are so low:

2. Similarly, in [line 276 of dataset.py](https://github.com/salmank255/ROAD_Waymo_Baseline/blob/65049d6e633c90cb5924b36a4c74f69e5551d491/data/datasets.py#L276), the ground truth bbox format is `(x1, y1, x2, y2)`. `g_w` should be multiplied by `x`, and `g_h` should be multiplied by `y`. However, you have multiplied them in reverse. This line should be rewritten as:
`box[bi] = box[bi]*g_w if bi % 2 == 0 else box[bi]*g_h`

3. In the mAP calculation method in `evaluation.py`, you can see in [line 427](https://github.com/salmank255/ROAD_Waymo_Baseline/blob/65049d6e633c90cb5924b36a4c74f69e5551d491/modules/evaluation.py#L417) that mAP is divided by all labels. However, in reality, it should only be divided by labels that have appeared in ground truth.

For example, as shown in the figure below:
(**this figure depicts the execution results after I fixed the bugs for 1. and 2.**)

![image](https://github.com/salmank255/ROAD_Waymo_Baseline/assets/62836757/9f23142f-805e-45a2-bf20-7cb9da451800)
If this video contains only `Class 1 `and `Class 5`, with `mAP` scores of `0.55` and `1` respectively, then the division should be by `2`, not by the total number of labels for all agents: `len(classes)`

I hope you can correct the above bugs. This should lead to an improvement in the mAP for all participants.

If necessary, you can also allow me to submit a merge request. Please reply below for further instructions.

**Comment by salmank255** [comment]
Hi @ricky-696 

Below are the responses to your points:

1. g_w and h_w are initialized to 0, however, they are global variables that we first update in the get_item function (https://github.com/salmank255/ROAD_Waymo_Baseline/blob/65049d6e633c90cb5924b36a4c74f69e5551d491/data/inference_dataset.py#L593) of the loader and then used in 'get_filtered_tubes' and 'get_filtered_frames function' while doing the evaluation.

2. We did a mistake in the first place while assigning values (https://github.com/salmank255/ROAD_Waymo_Baseline/blob/65049d6e633c90cb5924b36a4c74f69e5551d491/data/inference_dataset.py#L595) we assigned height to g_w and width to g_h, but it's not a bug but typo it won't effect the performance of the model. 

3. We are not evaluating the video by video but finding the mean of all the aps (sap) https://github.com/salmank255/ROAD_Waymo_Baseline/blob/65049d6e633c90cb5924b36a4c74f69e5551d491/modules/evaluation.py#L417 

This is outside the loop of videos, but if you think there's a bug can you please send a PR to us we will confirm and merge if there is actually a bug.

Many thanks for finding these but still I don't think they are bugs.

**Comment by ricky-696** [comment]
Hi @salmank255 

Here are the responses to your reply:

1. When I was tracing, I directly called the [`evaluate_tubes`](https://github.com/salmank255/ROAD_Waymo_Baseline/blob/65049d6e633c90cb5924b36a4c74f69e5551d491/modules/evaluation.py#L355) function (**based on the [build_eval_tubes_roadpp function called in main.py](https://github.com/salmank255/ROAD_Waymo_Baseline/blob/65049d6e633c90cb5924b36a4c74f69e5551d491/main.py#L342)**) which reads `road_waymo_trainval_v1.0.json` as input. 
In the [`evaluate_tubes`](https://github.com/salmank255/ROAD_Waymo_Baseline/blob/65049d6e633c90cb5924b36a4c74f69e5551d491/modules/evaluation.py#L355)  function, the `VideoDataset`class is not used, it [directly calls the `get_gt_tubes`](https://github.com/salmank255/ROAD_Waymo_Baseline/blob/65049d6e633c90cb5924b36a4c74f69e5551d491/modules/evaluation.py#L396C32-L396C44) function to obtain gt_tubes.
This is why the variables `g_w` and `g_h` were defined as `0`. Could you please confirm whether the `VideoDataset` class is used in the Evaluation Code on EvalAI? It's possible that not using this class is causing my score to be 0.

2. So, is the reversed formulation `box[bi] = box[bi]*g_w if bi % 2 == 0 else box[bi]*g_h` due to incorrect assignment of values?

3. You're right, I overlooked that I'm only tracing a single video, not the entire training set.

I hope you can help me verify whether the code on EvalAI correctly assigns values to `g_w` and `g_h`, I think this is main reason why I got 0 points.

Thank you!

**Comment by salmank255** [comment]
1. In the evaluation we support a fixed size which is parser.add_argument('--MIN_SIZE', default=600, or box[bi] = box[bi]*840 if bi % 2 == 0 else box[bi]*600 not using g_w and g_h.

2. It's not the reversed formulation but we initialised the variables wrong, in our evaluation script we are using the hard values to make it fair and accept only one size of output which is  box[bi] = box[bi]*840 if bi % 2 == 0 else box[bi]*600

You can also change these values with yourself for both training and testing. The baseline does support different sizes but for the challenge, we fixed it to one size to have a fair comparison.

I am not 100% sure why your results are 0 but I can check your predictions maybe there is something wrong with them.

**Comment by ricky-696** [comment]
Hi, @salmank255


From your response 2, I understand the reason for my results being `0`. My predictions are based on the native dimensions of the video (1920, 1280), which do not match the competition's specified size of `(840, 600)`. Could you please let me know **where in the competition's documentation the requirement to use the size `(840, 600)` for evaluation is stated?**

Alternatively, is it necessary for the **uploaded results to have normalized bounding boxes?** However, I noticed that the examples on EvalAI show results in absolute values.

I will try uploading results using the `(840, 600)` size later to see if it works correctly. Thank you.
